File: src/algorithms/arrays/merge-sorted-arrays/merge-sorted-arrays.test.ts
import { describe, it, expect } from "vitest";
import { mergeSortedArrays } from "./sources/merge-sorted-arrays.ts?fn";

describe("mergeSortedArrays", () => {
  it("merges two basic sorted arrays into one sorted array", () => {
    const result = mergeSortedArrays([1, 3, 5], [2, 4, 6]);
    expect(result).toEqual([1, 2, 3, 4, 5, 6]);
  });

  it("handles empty first array", () => {
    const result = mergeSortedArrays([], [1, 2, 3]);
    expect(result).toEqual([1, 2, 3]);
  });

  it("handles empty second array", () => {
    const result = mergeSortedArrays([1, 2, 3], []);
    expect(result).toEqual([1, 2, 3]);
  });

  it("handles both arrays empty", () => {
    const result = mergeSortedArrays([], []);
    expect(result).toEqual([]);
  });

  it("handles overlapping values correctly", () => {
    const result = mergeSortedArrays([1, 2, 4], [2, 3, 5]);
    expect(result).toEqual([1, 2, 2, 3, 4, 5]);
  });

  it("handles single element arrays", () => {
    const result = mergeSortedArrays([5], [3]);
    expect(result).toEqual([3, 5]);
  });

  it("handles arrays of different lengths", () => {
    const result = mergeSortedArrays([1, 10], [2, 3, 4, 5, 6]);
    expect(result).toEqual([1, 2, 3, 4, 5, 6, 10]);
  });

  it("handles default input from the algorithm definition", () => {
    const result = mergeSortedArrays([1, 3, 5, 7, 9], [2, 4, 6, 8, 10]);
    expect(result).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
  });
});

File: src/algorithms/arrays/two-pointer/merge-sorted-arrays/__tests__/MergeSortedArrays_test.cpp
#include "../sources/MergeSortedArrays.cpp"
#include <cassert>
#include <iostream>
#include <vector>

int main() {
    // Basic merge
    assert((mergeSortedArrays({1, 3, 5}, {2, 4, 6}) == std::vector<int>{1, 2, 3, 4, 5, 6}));

    // Empty first array
    assert((mergeSortedArrays({}, {1, 2, 3}) == std::vector<int>{1, 2, 3}));

    // Empty second array
    assert((mergeSortedArrays({1, 2, 3}, {}) == std::vector<int>{1, 2, 3}));

    // Both empty
    assert(mergeSortedArrays({}, {}).empty());

    // Overlapping values
    assert((mergeSortedArrays({1, 2, 4}, {2, 3, 5}) == std::vector<int>{1, 2, 2, 3, 4, 5}));

    // Single elements
    assert((mergeSortedArrays({5}, {3}) == std::vector<int>{3, 5}));

    // Default input
    assert((mergeSortedArrays({1, 3, 5, 7, 9}, {2, 4, 6, 8, 10}) == std::vector<int>{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}));

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
